triplea-game/triplea
# Support for moderator single-use-key

## Overview
Adds support and concept for a single-use-key that moderators can use to register a
password and be granted a permanent API key that is stored on their system.

This allows for keys to be distributed and then a new one to be regenerated that
is not known to anyone other than the moderator.

Change Summary:
- incorporate logic so that a key password is registered and hashed with the main moderator key
- update UI to prompt for single-use-key and password if no API key is stored in client settings
- Update toolbox flow to always at least ask for api key password
- add DB table for single-use-key
- update backend to do validation of api key salted with password

## Functional Changes
See change summary in overview


## Manual Testing Performed
- use the SQL posted in readme
- launch the main 'show toolbox controller'
- click 're-register' to clear out any stored API keys in client settings
- use api key value 'password' to register a new API key, password is of your choosing, I typically use 1234
- following registration enter 1234 in api key password prompt
- should then see the moderator toolbox

* server application needs to be running
* flyway needs to be run to get the latest DB code

To reset single use-key without re-running flyway, execute DB command:
```
update moderator_single_use_key set date_used = null;
```

One note, given the hashing and security of the API key, debugging is a bit of a PITA if things do not work right.

## Before & After Screen Shots



### After
![Screenshot from 2019-06-10 14-31-28](https://user-images.githubusercontent.com/12397753/59229188-9997f380-8b8e-11e9-8118-8da8b9b5fa8f.png)
![Screenshot from 2019-06-10 14-31-37](https://user-images.githubusercontent.com/12397753/59229189-9997f380-8b8e-11e9-90d5-0c0974111def.png)


## Additional Review Notes
- once we get the last of moderator toolbox merged up I intend to spend some time doing some polishing changes. Happy to have anything/everything get called out, though some updates will likely conflict with in-flight changes and I may opt to follow-up. 
- More changes on the way! There is another branch inflight that adds more tabs.
- Reasonable number of interesting logics introduced, notably to do key hashing.
- DB changes are listed at the end of the diff, it might be easiest to see those first and then jump back to the top.
